### DISCUSSION TITLE
Implicit conversion depth constraint for parameter

### DISCUSSION THREAD

**Discussion by SanftMonster** [dotnet/csharplang#6068] [Language Ideas]
Hello,

I'd like to share my idea about adding constraint to the implicit conversion depth of parameter. It tells the compiler the maximum steps that could be accepted to do implicit conversion for a parameter of a method.

For instance, I have a class below, which is a vector and is supposed to support implicit conversion from scalar to a vector of length 1.

```
public class Vector<T> where T : struct, IEquatable<T>, IConvertible
{
    public static implicit operator Vector<T>(T value) => new Vector<T>(value);
}
public class ArithExtension{
    public static Vector<double> Add(this Vector<double> lhs, Vector<int> rhs)
    {
        // implementation
    }
    public static Vector<double> Add(this Vector<double> lhs, Vector<double> rhs)
    {
        // implementation
    }
}
```
Then if I want to write some code like below, the compiler would tell me it is ambiguous.

```
Vector<double> v = new Vector(10);
v = v.Add(3);
```

That's because the type ```int``` could implicitly convert to ```double```, so the value "3" could be either converted to ```Vector<int>``` or ```Vector<double>```.

### Proposal
Therefore, I think all things will be good if I can add a constraint for such implicit conversion. The feature may look like the code below.

```
public class ArithExtension{
    public static Vector<double>(this Vector<double> lhs, Vector<int> rhs)
    {
        // implementation
    }
    public static Vector<double>(this Vector<double> lhs, [ImplicitConversionDepth(1)] Vector<double> rhs)
    {
        // implementation
    }
}
```

The parameter for the attribute is an integer which indicates the maximum steps of implicit conversion the parameter could accept. If it is set to 0, then no implicit conversion would be accepted.

### Other conditions
I'd like to share one of the other conditions that may benefit from this feature. Say an library that describes dog and robot dog is used by us. For some reason, the library was not well designed and there is an implicit conversion from robot dog to dog. Now, we want to design a farm which contains dogs and robot dogs as guards. Since we need to feed the dogs, we may write the code below.

```
public class Farm{
    public void Feed(Chicken chicken){}
    public void Feed(Dog dog){}
    ......
}
```
Then some bad things happened, robot dogs could also be fed, which is in fact unexpected. But with the feature, we can write the code like below.

```
public class Farm{
    public void Feed(Chicken chicken){}
    public void Feed([ImplicitConversionDepth(0)] Dog dog){}
    ......
}
```

That is, if the library we reference has some design of implicit conversion which we do not want. Adding a constraint to the maximum steps of implicit conversion may be a good way to deal with it.

**Comment by CyrusNajmabadi** [comment]
Hi there. What you are asking for today is already possible. Someone can already add such an attribute and then write an analyzer to error if code ends up using more implicit conversions than what is allowed. Thanks!

**Reply by svick** [reply]
@CyrusNajmabadi An analyzer won't help here. The problem is that [the desired code](https://sharplab.io/#v2:C4LgbgpgTglgZgTwD4DUIGNgHsoB4AmWArgEYA2EAfAARjUC81AdhAO7VqY4HHlUAUARgAMASgDcAWABQdRmAB0AQXz5+AZgkyZAAXXUdAJg4ZseACo1WAC2gRq56iGoBnYFCKYZAbxnV/BvqcZvyOYACGZEQQotTeAL7a0gGBBoIAbNQwALYADmQw6DDA1Fi50OFmJlwWlKG0kdGx9DQs7MHclvwRUTFS0onSuvo6GQbGSrDA1gCiAB7AEEwuMFhMPn4BemmZHXiEpBQ0KmrTMC7VZjyHVNRk1i4ANJfcMEzANFAPopv+vskpALTKBYdhMIhkMj9FKDFLbUa7UzcA58Y6qfhnC57a6ou4PZ7YlFHahfFw/AF/X6A4Gg5gQqFUwbxIA=) today produces:

> error CS0121: The call is ambiguous between the following methods or properties: 'ArithExtension.Add(Vector\<double>, Vector\<int>)' and 'ArithExtension.Add(Vector\<double>, Vector\<double>)'

The goal seems to be to avoid this compiler error, not to introduce a new error.

**Reply by SanftMonster** [reply]
> anks!

Hello, thank you for your answer! @svick is right, my goal is to let the compiler constraints the maximum times of implicit conversion, so that some compile errors or unexpected behaviors could be avoided without re-designing the APIs.
